### DISCUSSION TITLE
save settlers, commander into an xml/json file with experience

### DISCUSSION THREAD

**Discussion by shirishag75** [mars-sim/mars-sim#430] [Ideas]
is there or would there be a way to have a crew with experience which you can throw into any simulation. Something let's say like the crew of NCC-1701 or whatever fancies a person decides to name the crew. As shared in https://github.com/mars-sim/mars-sim/issues/212 it would not only just have or add continuity but would make it for interesting things/events to happen sooner than later.

**Comment by mokun** [comment]
@shirishag75 , 

Yes we can try make each settler save-able. 

@Urwumpe , can you help also ?

**Comment by mokun** [comment]
If a settler from mars-sim is a real human being, perhaps the most interesting part would be one's mind because it is supposed to carry the `experiences` of a lifetime of a person.

However, at this point, the skill points of a settler in a career is just a static and predictable number. 

But what I've been planning to do is to make use of the concept of the Machine Learning (ML) to store experiences. See [this](https://github.com/mars-sim/mars-sim/issues/183) 

The mind of a settler would simply be a `matrix` of `weights`.  By using these weights, you automatically get some *fuzziness* on how well a person can or cannot perform a task, etc. This mimics better the task capability of a person.

For now, may be we can start with a 7x7 matrix. 

It means 7 inputs (7 rows of 0 or 1), 7 hidden layers, and 1 output.

7 rows of 0 or 1 = 2 ^ 7 = 128 possible values.

Beside looking for the right activating function to do the job for us, we will need to map out exactly what each of each of the 128 values would point to. 

Once it's done, it will have a lot more motivation to make a persistent copy of a person since his experience can carry on !

**Comment by shirishag75** [comment]
There is another thing that I want people to keep in mind or hopefully something we can also think of in future, having school/s to teach specific tasks. Something which could be opened up once the initial survival bit is done or something on those lines perhaps. What would people think about that. There is also genetics or possibility of any born people to inherit genes, habits, skills from one or both parents.I dunno how do we get to know what a potential a child has. Let mw know what are your thoughts on that.

**Comment by mokun** [comment]
> There is also genetics or possibility of any born people to inherit genes, habits, skills from one or both parents.

We do have a simple [Gene](https://github.com/mars-sim/mars-sim/blob/master/mars-sim-core/src/main/java/org/mars_sim/msp/core/person/Gene.java) class getting ready to roll. 

![Gene](https://user-images.githubusercontent.com/1168584/63535165-ace80900-c4c5-11e9-9c43-13e8d17e81cc.jpg)

Nothing fancy at this point. We have things set up to record blood type, height, weight, and attributes of a person as that person is being created.

https://github.com/mars-sim/mars-sim/blob/4f6ae63562489ebffc5370850bb7b4455cfff3c0/mars-sim-core/src/main/java/org/mars_sim/msp/core/person/Person.java#L330-L511

We just did not have the chance of putting in more codes to stitch things together to make a particular use case of the genetics.

> I dunno how do we get to know what a potential a child has.

Yes. *Settling down* could mean having **the first baby on Mars**, which inaugurating a new branch of humanity in our history book. 

In this case, we could use that `Gene` class to create the baby that would inherit the traits of both parents.

> hopefully something we can also think of in future, having school/s to teach specific tasks. Something which could be opened up once the initial survival 

It's true that human being needs stimulus and learn a new trade is good for growing our brain. 

Even right now in the simulation, a settler can teach skill to another but it's done one-to-one and in a rather unpredictable manner. 

Definitely need to be more *methodical*, such as setting up class hour. 

e.g. we could come up with concepts of a model and an interface to allow players to define how skill/knowledge are passed on.

Also, I have thought of the checkbox option of having *mandatory* safety training because safety is a mindset that one needs to develop.

**Comment by shirishag75** [comment]
I would clearly love that as well as repairing skillset.  At least in the initial run/s it would be imperative to have production, repair, learning, expanding the base and opening up science. I like Schiaparelli Point as there are only four people and so we can do things without going crazy and observing things. I wish though there was a way in the headless version to only follow stuff at say Schiaparelli Point and once you have made some decent progress, you go to the other bases and help out . It would also make for some nice progression in terms of handling things. What do people think about that ?

**Reply by mokun** [reply]
See discussions on https://github.com/mars-sim/mars-sim/issues/183.

**Comment by mokun** [comment]
> once you have made some decent progress, you go to the other bases and help out 

We could come up with different *simulation scenarios* for different initial setup.

I'll open up a brand new thread to discuss scenarios.
